Portage of pipeline 1 dtbook cleaning routines

NPavie opened the pull request:
Reintagrating pipeline 1 dtbook cleaning routines in pipeline 2.

The routines logic is stored in dtbook-utils modules while it is exposed as a pipeline script with a new dtbook-cleaner script modules.

bertfrees commented on modules/scripts-utils/dtbook-utils/pom.xml:
```diff
@@ -64,7 +76,8 @@
     <expose-services>
-      org.daisy.pipeline.modules.impl.Module_dtbook_utils
+        org.daisy.pipeline.modules.impl.Module_dtbook_utils,
+        org.daisy.pipeline.modules.dtbook_utils.impl.DTBookCleanerLibrary$Provider
```
I will move this to org.daisy.pipeline.dtbook.saxon.impl.

bertfrees replied:
Fixed

bertfrees commented on modules/scripts-utils/dtbook-utils/src/main/resources/xml/dtbook-utils-library.xpl:
```diff
@@ -18,5 +18,7 @@
     <p:import href="dtbook-to-mods-meta.xpl"/>
     <p:import href="dtbook-update-links.xpl"/>
+
+    <p:import href="fix-dtbook/fix-dtbook.xpl"/>
```
I think I will move the dtbook-cleaner script inside the dtbook-utils module so that

- we don't need to create a new module that we will delete afterwards
- we don't need to expose the new steps to the library yet

For now the dtbook-cleaner script will not be actually exposed as a Pipeline script yet, but I will create a custom release with the script enabled for Save-as-DAISY.

bertfrees replied:
Removed from library.

bertfrees commented on modules/scripts-utils/dtbook-utils/src/main/resources/xml/fix-dtbook/fix-dtbook.xpl:
```diff
@@ -0,0 +1,111 @@
+                xmlns:xs="http://www.w3.org/2001/XMLSchema"
+                type="px:dtbook-fix" name="main">
+    <p:documentation xmlns="http://www.w3.org/1999/xhtml">
+        <h1 px:role="name">Fix DTBook XML</h1>
```
I know that this happens in other steps too, but these `px:role` attributes only really serve a purpose within script definitions. In a commit that I haven't pushed yet I've removed (most) `px:role` attributes from within the documentation of non-script XProc files.

bertfrees replied:
So this is better:

```xml
<h1>Fix DTBook XML</h1>
<p>Apply fixing routines on a given dtbook xml.</p>
```

bertfrees replied:
Fixed

bertfrees commented on modules/scripts-utils/dtbook-utils/src/main/resources/xml/fix-dtbook/fix-dtbook.xpl:
```diff
@@ -0,0 +1,111 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Can you include some documentation to explain that this code came from Pipeline 1 originally?

bertfrees replied:
@NPavie it would be nice to have a comment here that the module was ported from Pipeline 1, and with a mentioning of both you and the original authors.

bertfrees replied:
@NPavie I don't see Pipeline 1 and the original authors mentioned anywhere yet.

NPavie replied:
(sorry i just see now the comment you made in the review tab)

Noted, will do

NPavie replied:
I got back the list of authors that have participated to the dtbook fix script.
I will add a comment stating this script is a portage of pipeline 1 se_tpb_dtbookFix transformer followed by the list of authors in order of participation.

NPavie replied:
should be good in 59df9d0bd3f5fa95cfde4dbc632daaa5af352eed

NPavie replied:
I'll also add a similar mention with list of authors in other pipeline 1 ported scripts

bertfrees replied:
Thanks!

bertfrees commented on modules/scripts-utils/dtbook-utils/src/main/java/org/daisy/pipeline/modules/dtbook_utils/impl/DTBookCleanerLibrary.java:
```diff
@@ -0,0 +1,32 @@
+        }
+    }
+
+    public String getDefaultLocale(){
+        return Locale.getDefault().toString().replace('_', '-');
+    }
+
+    public String getDate(){
+        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
+        return sdf.format(new Date());
+    }
```
These can be changed to static methods, so that you don't need the `Library:new()` in library.xsl.

bertfrees replied:
Done

bertfrees commented on modules/scripts-utils/dtbook-utils/pom.xml:
```diff
@@ -23,6 +23,14 @@
+    <dependency>
+      <groupId>org.daisy.libs</groupId>
+      <artifactId>com.xmlcalabash</artifactId>
+      <version>1.1.20-p12-98</version>
```
Version should be updated in bom/pom.xml.

bertfrees replied:
Done

bertfrees commented on modules/scripts-utils/dtbook-utils/src/main/resources/xml/dtbook-utils-library.xpl:
```diff
@@ -20,5 +20,6 @@
     <p:import href="dtbook-update-links.xpl"/>
 
     <p:import href="fix-dtbook/fix-dtbook.xpl"/>
+    <p:import href="fix-dtbook/doctyping.xpl"/>
```
We already had functionality in Pipeline to set the doctype (with `px:fileset-store`). We're including the step it in this release but we're going to clean it up afterwards.

bertfrees commented on modules/scripts/dtbook-cleaner/src/main/resources/xml/dtbook-cleaner.script.xpl:
```diff
@@ -0,0 +1,164 @@
+    <p:documentation xmlns="http://www.w3.org/1999/xhtml">
+        <h1 px:role="name">DTBook Cleaner</h1>
+        <p px:role="desc">Cleaning DTBook documents</p>
+        <div px:role="author maintainer">
+            <p px:role="name">Nicolas Pavie</p>
+            <a px:role="contact" href="mailto:pavie.nicolas@gmail.com">pavie.nicolas@gmail.com</a>
+            <p px:role="organization">DAISY Consortium</p>
+        </div>
```
I will change this to:

```xml
<address>
    Authors:
    <dl px:role="author">
        <dt>Name:</dt>
        <dd px:role="name">Nicolas Pavie</dd>
        <dt>E-mail:</dt>
        <dd><a px:role="contact" href="mailto:pavie.nicolas@gmail.com">pavie.nicolas@gmail.com</a></dd>
        <dt>Organisation:</dt>
        <dd px:role="organization">DAISY Consortium</dd>
    </dl>
</address>
```

bertfrees replied:
Done

bertfrees commented on modules/scripts/dtbook-cleaner/src/main/resources/xml/dtbook-cleaner.script.xpl:
```diff
@@ -0,0 +1,164 @@
+            <p:with-option name="publisher" select="$publisher" />
+        </px:dtbook-fix>
+
+        <p:choose>
+            <p:when test="$WithDoctype">
+                <!-- dtbook with doctype (result is serialized) -->
+                <px:dtbook-doctyping/>
+            </p:when>
+            <p:otherwise>
+                <p:identity />
+            </p:otherwise>
+        </p:choose>
+        <!-- Store on disk -->
+        <p:store>
+            <p:with-option name="href" select="concat(resolve-uri($output-dir),$output-name)"/>
+        </p:store>
```
We'll replace this code later with `px:fileset-store`.

bertfrees commented on modules/scripts/dtbook-cleaner/src/main/resources/xml/dtbook-cleaner.script.xpl:
```diff
@@ -0,0 +1,164 @@
+
+
+    <!-- Options from dtbook-fix script -->
+    <p:option name="repair" required="false" select="false()" px:type="boolean">
```
Script option are always of type string. So the `select` attribute must become: `select="'false'"`. The `px:type="boolean"` is for telling the user interface what are the allowed values.

Note that this also means that in the `px:dtbook-fix` call below, in order to pass the options as booleans, the `p:with-option` must become: `<p:with-option name="repair" select="$repair='true'"/>`

bertfrees replied:
Fixed

bertfrees commented on modules/scripts/dtbook-cleaner/src/main/resources/META-INF/catalog.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<catalog xmlns="urn:oasis:names:tc:entity:xmlns:xml:catalog" xmlns:px="http://www.daisy.org/ns/pipeline">
+  <uri name="http://www.daisy.org/pipeline/modules/dtbook-cleaner/dtbook-cleaner.script.xpl"
```
I'm going to drop the `name`. (It is not required for a script.)

bertfrees replied:
Done

bertfrees commented on modules/scripts/dtbook-cleaner/pom.xml:
```diff
@@ -0,0 +1,71 @@
+
+  <properties>
+    <expose-services>
+      org.daisy.pipeline.modules.impl.Module_dtbook_cleaner,
```
I'm going to drop the `Module` implementation. (The Java class won't be generated if I drop the `name` of the script in catalog.xml.)

bertfrees reviewed (changes requested):
@NPavie Very nice job! Thank you.

Shall I wait for that more substantial test you were talking about?

NPavie commented:
@bertfrees sorry just saw your comment.

I don't think i'll have time to add the tests before the next release.
If you want to merge the PR for the upcoming release, i can do a new PR later to include them (but i need some time to prepare them).

bertfrees commented on modules/scripts/dtbook-cleaner/src/main/resources/xml/dtbook-cleaner.script.xpl:
```diff
@@ -0,0 +1,164 @@
+        </p:documentation>
+    </p:option>
+
+    <p:option name="ApplySentenceDetection" select="false()" px:type="boolean">
+        <p:documentation xmlns="http://www.w3.org/1999/xhtml">
+            <h2 px:role="name">Apply sentences detection</h2>
+            <p px:role="desc">Encapsulate sentences within the document</p>
+        </p:documentation>
+    </p:option>
```
@NPavie This option does not seem to be used?

NPavie replied:
Sorry i had not seen your messages : 
this was to be used for sentence detection in post-processing of dtbook for the addin (readded in last commit)

bertfrees commented on modules/scripts/dtbook-cleaner/src/main/resources/xml/dtbook-cleaner.script.xpl:
```diff
@@ -0,0 +1,164 @@
+    </p:import>
+    <p:import href="http://www.daisy.org/pipeline/modules/dtbook-break-detection/library.xpl">
+        <p:documentation>
+            px:dtbook-break-detect
+            px:dtbook-unwrap-words
```
@NPavie Are these also supposed to be used?

bertfrees commented on modules/scripts-utils/dtbook-utils/src/main/resources/xml/fix-dtbook/fix-dtbook.xpl:
```diff
@@ -0,0 +1,111 @@
+    <p:option name="fixCharset" required="false" select="false()">
+        <p:documentation xmlns="http://www.w3.org/1999/xhtml">
+            <h2 px:role="name">Repair - Fix Charset</h2>
+            <p px:role="desc">Fix the document charset (To be implemented)</p>
```
Is this option not implemented yet? Let's drop it from the script options then (comment it out).

bertfrees replied:
@NPavie Is this option going to be implemented or should we drop it?

bertfrees commented on modules/scripts-utils/dtbook-utils/src/main/resources/xml/fix-dtbook/fix-dtbook.xpl:
```diff
@@ -0,0 +1,111 @@
+            <p px:role="desc">Apply tidying routines on the dtbook</p>
+        </p:documentation>
+    </p:option>
+    <p:option name="simplifyHeadingLayout" required="false" select="false()">
+        <p:documentation xmlns="http://www.w3.org/1999/xhtml">
+            <h2 px:role="name">Tidy - Simplify headings layout</h2>
+            <p px:role="desc">TBD</p>
+        </p:documentation>
+    </p:option>
+    <p:option name="externalizeWhitespace" required="false" select="false()">
+        <p:documentation xmlns="http://www.w3.org/1999/xhtml">
+            <h2 px:role="name">Tidy - Externalize whitespaces</h2>
+            <p px:role="desc">TBD</p>
+        </p:documentation>
+    </p:option>
```
Are these also to be implemented?

bertfrees replied:
I've commented out the "TBD" descriptions for now.

bertfrees commented on modules/scripts-utils/dtbook-utils/src/test/xprocspec/test_fix.xprocspec:
```diff
@@ -0,0 +1,61 @@
+                    <head>
+                        <meta name="dc:Publisher" content="DAISY Consortium"/>
+                        <meta name="dc:Language" content="fr-FR"/>
+                        <meta name="dtb:uid" content="AUTO-UID-d990e3"/>
```
This is brittle. We can perhaps change it later with `http://www.daisy.org/pipeline/modules/common-utils/generate-id.xsl`.

bertfrees commented:
I've squashed your 4 commits, added a cleanup commit, and rebased onto the latest master.

bertfrees commented:
@NPavie Reminder that this PR needs some tests before it can be merged.

NPavie commented:
@bertfrees i finished a last round in tests on daisy #718 , i have a small update to make on SaveAsDAISY and i'll resume the creation of tests for this PR.

NPavie commented:
Hi @bertfrees, i added a small set of tests for the cleaning routines and a bugfix in an xsl that i detected while doing the tests.

bertfrees commented on modules/scripts-utils/dtbook-utils/src/main/resources/xml/fix-dtbook/narrator.xpl:
```diff
@@ -0,0 +1,47 @@
+        </p:documentation>
+    </p:option>
+
+    <p:xslt px:message="narrator-metadata">
```
@NPavie Do these messages need to show up in the job progress log?

NPavie replied:
(sorry i just see now the comment you made in the review tab)

Noted, will check that

NPavie replied:
Ok checked this, indeed the message does not need to be here anymore, i'll delete them.
(i think i added them for debugging at first)

NPavie replied:
@bertfrees should be good in 4c90bce9cbc334269111c20d77cf7a8c7445176c
(also removed those messages in the other xpls)

bertfrees replied:
Thanks!

bertfrees commented:
@NPavie Thanks for the fixes! Looking much better now. The only issues left are the following:

- One test seems to be failing
- The CLI is having trouble processing the option descriptions that you added. First, the problem was that the description didn't come through at all because the content of the `role='desc'` element is expected to be plain text containing valid markdown. So I changed that in commits 6c56e55 and be30b2d. Then, it turned out that the CLI does not play nice with HTML blocks, so I tried to fix that in commit 3ab8977. But still the output isn't very nice, I guess because the markdown handling of the CLI is so basic. Now my question is: does the description of these options need to be so extensive anyway? Does the end user need to know such details, or is it more aimed at the developer? In case of the former, my suggestion is to move the details to a separate file and link to it, similar to how it's done e.g. in the [dtbook-to-zedai docs](http://daisy.github.io/pipeline/modules/dtbook-to-zedai/doc/). In case of the latter, my suggestion is to move it to the `pxi:dtbook-fix` step instead. (Note that the documentation of regular steps is HTML, not Markdown.)
- The references to Pipeline 1, e.g. "Prepare dtbook for Pipeline 1 narrator": we should try to get rid of these.

NPavie commented:
@bert i fixed the failing test (forgot to change a title)

I also removed the mention to pipeline 1 narrator in the current documentation, to specify that it's more of a cleanup of the dtbook for audio-synthesis.

IMO, the documentation might be of use for users who would want to know more precisely what the script is doing while cleaning.
A separate doc folder/markdown file, with links to it in the script inner documentation, makes sense.
If its ok with you, i'll move the detailed script documentation to a `doc/index.md` following the dtbook-to-zedai model you mentionned.

Just in case i forget (but I might make a separate issue for it), another test is failing on windows for this module : 
```
Failed tests:
  test_load
    if the base uri starts with file:///
      * FAILURE: this results in a xml:base with only one slash after file:
```
On my machine, the resulting fileset as `xml:base="file:/F:/"` (F: is the disk my terminal is currently positionned on when running the tests)
As the "/" path does not really exists on Windows, it seems like the produced result is correct (but on windows case, the result itself will depends on dev system configuration and from which disk the test is run)
I don't know if you can disable tests on a system basis with XProcSpec  (like excluding a system for specific tests)

bertfrees commented:
@NPavie Thanks! Yes, moving the detailed script documentation to a doc/index.md is perfect.

Regarding the other failing test: yes, please create a new issue. That's something we want to fix but not in the upcoming release. You can not disable an XProcSpec scenarion based on which platform the tests are running on. If you want to do that you need to put it in a separate XProcSpec file and then mess with src/test/java/XProcSpecTest.java (override the `runXProcSpec` method of `AbstractXSpecAndXProcSpecTest`). But we should avoid that if possible. It's probably better to change the test. For now you could add a `pending` attribute.

bertfrees commented:
Merged in [ed88343](https://github.com/daisy/pipeline-modules/commit/ed88343f4d09ebfe07e2d7e92a3dae54ec4d8988)
